library/Fission/IPFS/Peer.hs

```diff
@@ -54,5 +62,45 @@ connect peer@(Peer peerID) = IPFSProc.run ["swarm", "connect"] (UTF8.textToLazyB
   (ExitFailure _ , _, _) -> Left $ CannotConnect peer
   (ExitSuccess   , _, _) -> Right ()
 
+peerAddressRe :: Regex
+peerAddressRe = mkRegex "^/ip[46]/([a-zA-Z0-9.:]*)/"
+
+extractIPfromPeerAddress :: String -> Maybe String
+extractIPfromPeerAddress peer = peer
+  & matchRegex peerAddressRe
+  <&> Data.List.head
```

Comment:
This `head :: [a] -> a` function is one of the reasons that people have started using custom Preludes. The problem with `head` is that it's _partial_: it's `undefined` for `head []`. `RIO` has this function under `RIO.Partial`, which is essentially guilt-tripping you into not using it unless you're _sure_ that you won't run into it.

Since we're arleady in `Maybe`, we can rewrite this like in my comment below on `isExternalIPv4`:

```haskell
extractIPfromPeerAddress :: String -> Maybe String
extractIPfromPeerAddress peer = do
  matchedString <- matchRegex peerAddressRgx peer
  headMaybe matchedString

-- or shorter, if you're comfortable with >>= --

extractIPfromPeerAddress :: String -> Maybe String
extractIPfromPeerAddress peer = matchRegex peerAddressRgx peer >>= headMaybe

-- less idiomatic, but closer to the original version --

extractIPfromPeerAddress :: String -> Maybe String
extractIPfromPeerAddress peer = peer & matchRegex peerAddressRgx >>= headMaybe
```

Reply:
> To help build the intuition of how <&> relates to >>=, notice that by switching to >>=, we stay on the happy path and get to use the inner "unwrapped" values. >>= is a superpowered &, where its superpower is keeping you on the happy path (only dealing with the Just values as if they weren't wrapped in a Just).

Thank you! Explicitly I knew that from functors in pictures (see plunger) however in practise I had forgot about it when constructing that pipe from earlier.

Anyhow
```haskell
extractIPfromPeerAddress :: String -> Maybe String
extractIPfromPeerAddress peer = matchRegex peerAddressRgx peer >>= headMaybe
```

reads the best imo